src/y_text.rs

```diff
@@ -56,30 +60,55 @@ impl YText {
         }
     }
 
+    /// Returns an underlying shared string stored in this data type.
+    pub fn __str__(&self) -> String {
+        match &self.0 {
+            SharedType::Integrated(v) => v.to_string(),
+            SharedType::Prelim(v) => v.clone(),
+        }
+    }
+
+    pub fn __repr__(&self) -> String {
+        format!("YText({})", self.__str__())
+    }
+
     /// Returns length of an underlying string stored in this `YText` instance,
     /// understood as a number of UTF-8 encoded bytes.
-    #[getter]
-    pub fn length(&self) -> u32 {
+    pub fn __len__(&self) -> usize {
         match &self.0 {
-            SharedType::Integrated(v) => v.len(),
-            SharedType::Prelim(v) => v.len() as u32,
+            SharedType::Integrated(v) => v.len() as usize,
+            SharedType::Prelim(v) => v.len(),
         }
     }
 
-    /// Returns an underlying shared string stored in this data type.
-    pub fn to_string(&self, txn: &YTransaction) -> String {
+    pub fn __iter__(&self) -> YTextIterator {
+        // Unsafe iterator references possible because blocks holding pointer results aren't ever deallocated.
+        let iterator = match &self.0 {
+            SharedType::Integrated(text) => unsafe {
+                let text_str = text.to_string();
+                let str_ptr: *const String = &text_str;
+                InnerYTextIter::Integrated((text_str, (*str_ptr).chars()))
+            },
+            SharedType::Prelim(string) => unsafe {
+                let str_ptr: *const String = string;
+                let iter: Chars<'static> = (*str_ptr).chars();
+                InnerYTextIter::Prelim(iter)
+            },
+        };
+
+        YTextIterator(ManuallyDrop::new(iterator))
+    }
+
+    pub fn __contains__(&self, pattern: String) -> bool {
         match &self.0 {
-            SharedType::Integrated(v) => v.to_string(txn),
-            SharedType::Prelim(v) => v.clone(),
+            SharedType::Integrated(_) => self.__str__().contains(&pattern),
+            SharedType::Prelim(string) => string.contains(&pattern),
         }
     }
 
     /// Returns an underlying shared string stored in this data type.
-    pub fn to_json(&self, txn: &YTransaction) -> String {
-        match &self.0 {
-            SharedType::Integrated(v) => v.to_string(txn),
-            SharedType::Prelim(v) => v.clone(),
-        }
+    pub fn to_json(&self) -> String {
+        self.__str__()
```

Comment:
`to_json` and `__str__` are not exactly the same - JSON strings are quoted between `"` characters. Besides that all `"` characters within JSON string must be escaped. While atm. there's no built-in method for that, you can achieve this by using:

```rust
let mut json_string = String::new();
Any::String(ytext.to_string().into_boxed_str()).to_json(&mut json_string);
json_string
```

Reply:
All json strings should be wrapped in double quotes? Does this test case look correct?
```python
y_text = YText("Hello World!")
...
assert y_text.to_json() == '"Hello World!"'
```

Reply:
Yeah, this looks super weird, but I think @Horusiath is right on this one.. 

The main reason is that other data types need to be able to call `to_json()` on child elements and receive a valid json string. E.g. Assume `ymap = ydoc.get_map(); ymap.set('text', Y.Text())` . Now the `to_json()` method should return a valid JSON string. `ymap.to_json() == '{ "text": "" }'`. This is only possible if the child element return double quotes.

An alternative solution would be to return data types instead of a json string. I.e. `ymap.to_json() == { "text": "Hello World" }` (note that this is now an object). This is what I'm doing in JavaScript since json is native to JavaScript. We could call this method differently in ypy to match the language but provide the same functionality.